File: lab/javascript/arrays_e_listas/maximo_subarray.js
function maximoSubarray(lista) {
  let somaAtual = lista[0];
  let somaMaxima = lista[0];

  for (let i = 1; i < lista.length; i++) {
    const x = lista[i];

    if (somaAtual + x < x) {
      somaAtual = x;        // recomeça soma
    } else {
      somaAtual = somaAtual + x;  // continua
    }

    if (somaAtual > somaMaxima) {
      somaMaxima = somaAtual;
    }
  }

  return somaMaxima;
}

function maximoSubarrayIndices(lista) {
  let somaAtual = lista[0];
  let somaMaxima = lista[0];
  let iniAtual = 0;
  let iniMax = 0;
  let fimMax = 0;

  for (let i = 1; i < lista.length; i++) {
    const x = lista[i];

    if (somaAtual + x < x) {
      somaAtual = x;
      iniAtual = i;
    } else {
      somaAtual = somaAtual + x;
    }

    if (somaAtual > somaMaxima) {
      somaMaxima = somaAtual;
      iniMax = iniAtual;
      fimMax = i;
    }
  }

  return [somaMaxima, [iniMax, fimMax]];
}


console.log(maximoSubarray([-2, 1, -3, 4, -1, 2, 1, -5, 4]));  // 6
console.log(maximoSubarrayIndices([-2, 1, -3, 4, -1, 2, 1, -5, 4]));  // [6, [3, 6]]

File: lab/python/arrays_e_listas/maximo_subarray.py
def maximo_subarray(lista):
    # supõe lista não vazia
    soma_atual = lista[0]
    soma_maxima = lista[0]

    for i in range(1, len(lista)):
        x = lista[i]

        # soma_atual = max(x, soma_atual + x)
        if soma_atual + x < x:
            soma_atual = x      # recomeça em x
        else:
            soma_atual = soma_atual + x  # continua somando

        # soma_maxima = max(soma_maxima, soma_atual)
        if soma_atual > soma_maxima:
            soma_maxima = soma_atual

    return soma_maxima


def maximo_subarray_indices(lista):
    soma_atual = lista[0]
    soma_maxima = lista[0]
    ini_atual = 0
    ini_max = 0
    fim_max = 0

    for i in range(1, len(lista)):
        x = lista[i]

        if soma_atual + x < x:
            soma_atual = x
            ini_atual = i
        else:
            soma_atual = soma_atual + x

        if soma_atual > soma_maxima:
            soma_maxima = soma_atual
            ini_max = ini_atual
            fim_max = i

    return soma_maxima, (ini_max, fim_max)  # soma e intervalo [ini..fim]


print(maximo_subarray([-2, 1, -3, 4, -1, 2, 1, -5, 4]))  # 6
print(maximo_subarray_indices([-2, 1, -3, 4, -1, 2, 1, -5, 4]))  # (6, (3, 6))
